service/remotecfg/remotecfg.go

```diff
@@ -0,0 +1,318 @@
+}
+
+// ServiceName defines the name used for the remotecfg service.
+const ServiceName = "remotecfg"
+
+// Options are used to configure the remotecfg service. Options are
+// constant for the lifetime of the remotecfg service.
+type Options struct {
+	Logger      log.Logger // Where to send logs.
+	StoragePath string     // Where to cache configuration on-disk.
+}
+
+// Arguments holds runtime settings for the remotecfg service.
+type Arguments struct {
+	URL              string                   `river:"url,attr,optional"`
+	ID               string                   `river:"id,attr,optional"`
+	Metadata         map[string]string        `river:"metadata,attr,optional"`
+	PollFrequency    time.Duration            `river:"poll_frequency,attr,optional"`
+	HTTPClientConfig *config.HTTPClientConfig `river:",squash"`
+}
+
+// GetDefaultArguments populates the default values for the Arguments struct.
+func GetDefaultArguments() Arguments {
+	return Arguments{
+		ID:               agentseed.Get().UID,
+		Metadata:         make(map[string]string),
+		PollFrequency:    1 * time.Minute,
+		HTTPClientConfig: config.CloneDefaultHTTPClientConfig(),
+	}
+}
+
+// SetToDefault implements river.Defaulter.
+func (a *Arguments) SetToDefault() {
+	*a = GetDefaultArguments()
+}
+
+// Validate implements river.Validator.
+func (a *Arguments) Validate() error {
+	// We must explicitly Validate because HTTPClientConfig is squashed and it
+	// won't run otherwise
+	if a.HTTPClientConfig != nil {
+		return a.HTTPClientConfig.Validate()
+	}
+
+	return nil
+}
+
+// Hash marshals the Arguments and returns a hash representation.
+func (a *Arguments) Hash() (string, error) {
+	b, err := river.Marshal(a)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal arguments: %w", err)
+	}
+	return getHash(b), nil
+}
+
+// New returns a new instance of the remotecfg service.
+func New(opts Options) (*Service, error) {
+	basePath := filepath.Join(opts.StoragePath, ServiceName)
+	err := os.MkdirAll(basePath, 0750)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		opts:   opts,
+		ticker: time.NewTicker(math.MaxInt64),
+	}, nil
+}
+
+// Data is a no-op for the remotecfg service.
+func (s *Service) Data() any {
+	return nil
+}
+
+// Definition returns the definition of the remotecfg service.
+func (s *Service) Definition() service.Definition {
+	return service.Definition{
+		Name:       ServiceName,
+		ConfigType: Arguments{},
+		DependsOn:  nil, // remotecfg has no dependencies.
+	}
+}
+
+var _ service.Service = (*Service)(nil)
+
+// Run implements [service.Service] and starts the remotecfg service. It will
+// run until the provided context is canceled or there is a fatal error.
+func (s *Service) Run(ctx context.Context, host service.Host) error {
+	s.ctrl = host.NewController(ServiceName)
+
+	s.fetch()
+
+	// Run the service's own controller.
+	go func() {
+		s.ctrl.Run(ctx)
+	}()
+
+	for {
+		select {
+		case <-s.ch:
+			b, err := s.getAPIConfig()
```

Comment:
Is there a way we can make s.fetch() robust enough to be used here in addition to the other 2 references? For example, the call to s.fetch() in this func does not read/write a cache file. In practice probably not a huge deal on startup but those little nuances could be tough to track down in the future.

Reply:
I've specifically avoided falling back to a cache here; the flow controller should be rejecting any invalid config coming its way, so we don't need to fall back on anything here.

Do you believe it would make this easier to read if we extracted the flushing to disk in another helper or added an option on fetch to conditionally flush to disk instead?

Reply:
We've talked offline and reworked the logic into smaller, reusable pieces between fetching remotely and locally only; I personally think it does look cleaner that way. Thanks @erikbaranowski !